Format specifier fixes

anatoli26 opened the pull request:
@elliefm, @obache, all,

I tried to pay a lot of attention to the details, but please check this PR exhaustively, each change separately, as there are many small details with a single char changing the entire meaning.

@elliefm, also please test-run this PR on as much platforms and OS as possible.

On the other hand, I have some observations:
 * please check the commit msg of the 1st commit in the PR. The `configure.am` checks don't make much sense for a _real_ detection of the specifier, but I proceeded as was suggested by @elliefm.
 * **imap/mailbox.c**, **imap/sync_support.c**, etc.: it was `%lu`, now it is `%ld`/`%lld`, i.e. signed/unsigned change. Not sure why it was unsigned as `time_t` is a signed data type.
 * **imap/mbexamine.c** line 259: why was that cast instead of `time_t`?
 * **master/master.c**: is `PRIu64` the correct specifier?

obache commented on configure.ac:
```diff
@@ -2333,16 +2333,26 @@ struct sockaddr_storage {
 
+#if (SIZEOF_TIME_T == SIZEOF_LONG_LONG_INT)
+#define TIME_T_FMT "%lld"
+#define strtoofft(nptr, endptr, base) strtoll(nptr, endptr, base)
```
should not be defined here `time_t` case , defined above `off_t` case.

obache replied:
or define `strtotimet` and change to use it for scanning `time_t`, for example, `parse_zoneinfo`
 https://github.com/cyrusimap/cyrus-imapd/blob/95cc49f06b9794012ddba8879f1a842374281b9a/imap/zoneinfo_db.c#L140
(already fixed for `zoneinfo_store` case)

anatoli26 replied:
@obache thanks! Fixed by removing the duplicate lines (I've altered the 1st commit and force-pushed it).

obache replied:
How about `strtol` usage?

anatoli26 replied:
@elliefm, what about the `strtoofft` function define that got removed with your 2 commits? Was it unused? And what about the `strtol` that @obache mentions?

elliefm replied:
Oh, I missed that hiding in there!  But searching the source tree, looks like it's never used.  Checking history, it was added in 2008, only used in one place, and then stopped being used in 2010.  So it's just cruft, and doesn't hurt to remove it.

I'm not sure about that `strtol` yet, will think about it some more...

elliefm commented on lib/auth_pts.c:
```diff
@@ -400,7 +401,7 @@ static int ptload(const char *identifier, struct auth_state **state)
         syslog(LOG_DEBUG,
                "ptload(): fetched cache record (%s)" \
-               "(mark %ld, current %ld, limit %ld)", identifier,
+               "(mark " TIME_T_FMT ", current %" PRIu64 ", limit %" PRIu64 ")", identifier,
```
I think all three of these should be TIME_T_FMT, not PRIu64? Note that at line 398/399 `now` is declared as `time_t`, and the printout of "current" and "limit" are calculated from `now`.

elliefm replied:
Yeah, they do need to be TIME_T_FMT.  Build fails on linux/gcc/x86_64 with PRIu64's here.

anatoli26 replied:
Right, got symbols-resolved var `now` to function `now()` which is `int64_t`. Will fix this on the next fixing iteration.

elliefm commented on imap/http_ischedule.c:
```diff
@@ -222,7 +222,7 @@ static int meth_get_isched(struct transaction_t *txn,
        the number of available RSCALEs and the config file size/mtime */
     assert(!buf_len(&txn->buf));
-    buf_printf(&txn->buf, "%ld-%zu-%ld-%ld", (long) compile_time,
+    buf_printf(&txn->buf, TIME_T_FMT "-%zu-" TIME_T_FMT "-" OFF_T_FMT, compile_time,
```
The %zu should probably also be changed to SIZE_T_FMT here

anatoli26 replied:
OK

elliefm commented on imap/caldav_alarm.c:
```diff
@@ -632,7 +632,7 @@ static int read_lastalarm(struct mailbox *mailbox,
 
     if (annot_buf.len &&
-        sscanf(buf_cstring(&annot_buf), "%ld %ld",
+        sscanf(buf_cstring(&annot_buf), TIME_T_FMT " " TIME_T_FMT,
```
Hmmm, does it make sense to use these format specifiers for `sscanf` conversions?  I'm not sure.  @ksmurchison, any thoughts?

anatoli26 replied:
Clang on OpenBSD was complaining:

```
imap/caldav_alarm.c:639:16: warning: format specifies type 'long *' but the argument has type
   'time_t *' (aka 'long long *') [-Wformat]
               &data->lastrun, &data->nextcheck) == 2) {
               ^~~~~~~~~~~~~~
imap/caldav_alarm.c:639:32: warning: format specifies type 'long *' but the argument has type
   'time_t *' (aka 'long long *') [-Wformat]
               &data->lastrun, &data->nextcheck) == 2) {                                                                                                                                                                                                                                                                
                               ^~~~~~~~~~~~~~~~
```

elliefm commented on imap/http_admin.c:
```diff
@@ -602,7 +602,7 @@ static int action_proc(struct transaction_t *txn)
 
         while (fgets(buf, sizeof(buf), f)) {
-            if (sscanf(buf, "btime %ld\n", &boot_time) == 1) break;
+            if (sscanf(buf, "btime " TIME_T_FMT "\n", &boot_time) == 1) break;
```
(ditto re `sscanf`)

anatoli26 replied:
The clang on OpenBSD was complaining:

```
imap/http_admin.c:605:44: warning: format specifies type 'long *' but the argument has type
  'time_t *' (aka 'long long *') [-Wformat]
            if (sscanf(buf, "btime %ld\n", &boot_time) == 1) break;
                                   ~~~     ^~~~~~~~~~
                                   %lld
```

elliefm reviewed (commented):
This looks okay to me by eye.  I'm about to do a local build and test run and see how that goes, will report back in a little bit.

elliefm commented:
@guimard Is there any way we could try this out on the Debian build farm, to see what pops up on the various platforms it tests?

elliefm commented on configure.ac:
```diff
@@ -2333,16 +2333,24 @@ struct sockaddr_storage {
-#define OFF_T_FMT "%ld"
-#define strtoofft(nptr, endptr, base) strtol(nptr, endptr, base)
-#elif (SIZEOF_OFF_T == SIZEOF_LONG_LONG_INT)
+#if (SIZEOF_OFF_T == SIZEOF_LONG_LONG_INT)
 #define OFF_T_FMT "%lld"
 #define strtoofft(nptr, endptr, base) strtoll(nptr, endptr, base)
+#elif (SIZEOF_OFF_T == SIZEOF_LONG)
+#define OFF_T_FMT "%ld"
+#define strtoofft(nptr, endptr, base) strtol(nptr, endptr, base)
 #else
 #error dont know what to use for OFF_T_FMT
 #endif
 
+#if (SIZEOF_TIME_T == SIZEOF_LONG_LONG_INT)
+#define TIME_T_FMT "%lld"
+#elif (SIZEOF_TIME_T == SIZEOF_LONG)
+#define TIME_T_FMT "%ld"
+#else
+#error dont know what to use for TIME_T_FMT
+#endif
+
```
Annoyingly, changing the order here breaks the linux/gcc/x86_64 build!

I think because on this platform, SIZEOF_OFF_T == SIZEOF_LONG == SIZEOF_LONG_LONG_INT, so whichever of those conditions we put first is the one that matches, but it _really_ wants `%ld`, and complains if it gets `%lld`, even though they're the same size...

elliefm replied:
(Same applies for TIME_T_FMT too: it really wants %ld, but all three types are the same size, so whichever is checked first wins.)

anatoli26 replied:
In this case we need to properly test the format specifiers in `configure.am`. I guess what you were trying with the redefined-type test is the way to go.

guimard commented:
@elliefm : of course we can

guimard commented:
@elliefm: test this patch or wait for some changes ?

elliefm commented:
Wait for some changes I think, but I thought I'd rope you into the discussion early. :)

anatoli26 commented:
> Wait for some changes I think, but I thought I'd rope you into the discussion early. :)

Please let me know how would you like to proceed. The `PRIu64` changes were based on what we discussed in the related issue. I could change the things you've marked, rebase and force-push the PR. Or would you prefer some other way?

How your redefined-type test should be incorporated in all this? The 1st commit should be deleted all-together or somehow replaced with your code.

anatoli26 commented:
I guess I could drop the 1st commit, incorporate the 2 changes (`PRIu64` & `%zu`) and it would be ready to incorporate your redefined-type check. OK?

guimard commented:
@anatoli26, @elliefm: to launch the cross platform test, I just need a patch for 3.2.4

elliefm commented:
> I guess I could drop the 1st commit, incorporate the 2 changes (`PRIu64` & `%zu`) and it would be ready to incorporate your redefined-type check. OK?

I need to fiddle more with the redefined-type check first, but I think this sounds like a plan.  Eventually, you should be able to add my fork as a remote, pull my branch, rebase your branch on top of it, and then force push it here, and my commit(s) with the redefined-type checks will end up as part of this PR too.

> to launch the cross platform test, I just need a patch for 3.2.4

Great, thanks.  Once we've got this working for at least both @anatoli26's system and mine, I'll make a branch/patch that is the same as this PR but based on 3.2 instead of master, and ping you to test it. :)

elliefm commented:
> I need to fiddle more with the redefined-type check first, but I think this sounds like a plan. Eventually, you should be able to add my fork as a remote, pull my branch, rebase your branch on top of it, and then force push it here, and my commit(s) with the redefined-type checks will end up as part of this PR too.

Okay, fiddling done, I think: https://github.com/cyrusimap/cyrus-imapd/compare/master...elliefm:v33/3128-time-offset-format-detection

So you should be able to rebase on top of this now, dropping your first commit in the process.  Please let me know if you need help. :)

Please check it on your system too -- it works fine for me, and I have no reason to expect it to not be portable, but let's not assume!

anatoli26 commented:
@elliefm,

I've rebased my branch on yours, dropped the 1st commit and made the 2 fixes.

On my OpenBSD it builds correctly now (i.e. no format specifier warnings) and `./configure` shows the following:
```
checking underlying integer type of `time_t'... long long int                                                                                                                                                                                                                                                           
checking printf format string for `time_t'... %lld                                                                                                                                                                                                                                                                      
checking underlying integer type of `size_t'... unsigned long int                                                                                                                                                                                                                                                       
checking printf format string for `size_t'... %lu                                                                                                                                                                                                                                                                       
checking underlying integer type of `off_t'... long long int                                                                                                                                                                                                                                                            
checking printf format string for `off_t'... %lld          
```

I guess we're ready to test it on other platforms.

anatoli26 commented:
@obache you mentioned something about `strtol` usage, but I can't find your comment here (got it via mail notifications).

Please let us know if you see something missing.

obache commented:
> @obache you mentioned something about `strtol` usage, but I can't find your comment here (got it via mail notifications).
> 
> Please let us know if you see something missing.

https://github.com/cyrusimap/cyrus-imapd/pull/3262#discussion_r510586635

elliefm commented:
I've updated my branch to do a similar thing to figure out the correct strtol-like parse function for these types, and defines `strtotimet`, `strtosizet` and (brings back) `strtoofft` based on it.  Also converted the `strtol` in zoneinfo_db.c that @obache found to use `strtotimet` instead.

You should be able to pull and rebase on top of it again to get the changes.

anatoli26 commented:
@elliefm, just rebased and fpushed everything. Should be good now, I've just compiled cyrus on OpenBSD with the composite patch and it looks ok, but please check everything again.

I had to change the way I merged the `%zu` fix before, it entered into a wrong commit (the http one), now it's in the correct one (the compile_time one), so all 4 format specifier changes of that line are in the same commit.

elliefm commented:
Great, thanks!

@guimard Here are the same commits as this PR, but on top of 3.2 instead of master: https://github.com/cyrusimap/cyrus-imapd/compare/cyrus-imapd-3.2...elliefm:v32/3262-patch

or as a single .patch file:
https://github.com/cyrusimap/cyrus-imapd/compare/cyrus-imapd-3.2...elliefm:v32/3262-patch.patch

Can you kick this off on the Debian farm please?  I'm particularly curious to see what happens on the platforms where `time_t` is only 32 bits, since that's been a nuisance in the past!

guimard commented:
Hi,

I launched the job, you'll see result in a few hours [here](https://buildd.debian.org/status/logs.php?pkg=cyrus-imapd&ver=3.2.4-5), and a few hours later, [test results on amd64, arm64, armhf and i386 only](https://release.debian.org/britney/pseudo-excuses-experimental.html#cyrus-imapd) _(wait for 3.2.4-5, nothing wrong means OK)_

guimard commented:
@elliefm: build succeeded for all architectures \o/

anatoli26 commented:
> @elliefm: build succeeded for all architectures \o/

But... there are some warnings related to this PR on i386:
```
master/master.c:1990:27: warning: format ‘%llu’ expects argument of type ‘long long unsigned int’, but argument 3 has type ‘rlim_t’ {aka ‘long unsigned int’} [-Wformat=]
 1990 |         syslog(LOG_DEBUG, "set maximum file descriptors to %" PRIu64 "/%" PRIu64,
      |                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1991 |                rl.rlim_cur, rl.rlim_max);
      |                ~~~~~~~~~~~ 
      |                  |
      |                  rlim_t {aka long unsigned int}
In file included from master/master.c:73:
/usr/include/inttypes.h:105:34: note: format string is defined here
  105 | # define PRIu64  __PRI64_PREFIX "u"
master/master.c:1990:27: warning: format ‘%llu’ expects argument of type ‘long long unsigned int’, but argument 4 has type ‘rlim_t’ {aka ‘long unsigned int’} [-Wformat=]
 1990 |         syslog(LOG_DEBUG, "set maximum file descriptors to %" PRIu64 "/%" PRIu64,
      |                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1991 |                rl.rlim_cur, rl.rlim_max);
      |                             ~~~~~~~~~~~
      |                               |
      |                               rlim_t {aka long unsigned int}
In file included from master/master.c:73:
/usr/include/inttypes.h:105:34: note: format string is defined here
  105 | # define PRIu64  __PRI64_PREFIX "u"
master/master.c:1996:16: warning: format ‘%llu’ expects argument of type ‘long long unsigned int’, but argument 3 has type ‘rlim_t’ {aka ‘long unsigned int’} [-Wformat=]
 1996 |                "setrlimit: Unable to set file descriptors limit to %" PRIu64 ": %m",
      |                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1997 |                rl.rlim_cur);
      |                ~~~~~~~~~~~
      |                  |
      |                  rlim_t {aka long unsigned int}
In file included from master/master.c:73:
/usr/include/inttypes.h:105:34: note: format string is defined here
  105 | # define PRIu64  __PRI64_PREFIX "u"
```
Looks like we can't use `PRIu64` directly and `rlim_t` should also have its own format specifier.


```
imap/index.c: In function ‘extract_attachment’:
imap/index.c:5499:21: warning: format ‘%ld’ expects argument of type ‘long int’, but argument 13 has type ‘size_t’ {aka ‘unsigned int’} [-Wformat=]
 5499 |                     "PUT %s/%s %s\r\n"
      |                     ^~~~~~~~~~~~~~~~~~
......
 5510 |                     type, subtype, buf_cstring(&buf), buf_len(data));
      |                                                       ~~~~~~~~~~~~~
      |                                                       |
      |                                                       size_t {aka unsigned int}
imap/index.c:5506:40: note: format string is defined here
 5506 |                     "Content-Length: %ld\r\n"
      |                                      ~~^
      |                                        |
      |                                        long int
      |                                      %d
```

And it looks like we've missed some hardcoded format specifiers as `buf_len` resolves to `size_t`:
```
lib/util.h
307:size_t buf_len(const struct buf *buf);
```

```
imap/mailbox.c: In function ‘_commit_one’:
imap/mailbox.c:1878:69: warning: comparison of integer expressions of different signedness: ‘ssize_t’ {aka ‘int’} and ‘uint32_t’ {aka ‘unsigned int’} [-Wsign-compare]
 1878 |     if (retry_write(mailbox->index_fd, buf, mailbox->i.record_size) != mailbox->i.record_size) {
      |                                                                     ^~
```
And there are some other unrelated, though probably important warnings to consider. In total only in the i386 build there are more than 300 warnings. I believe we have to check them all to be sure everything works as expected. Different signedness sometimes may lead to security issues.

anatoli26 commented:
> So it looks like we can't use `PRIu64` directly and `rlim_t` should also have its own format specifier.
> 
> And I guess we've missed some hardcoded format specifiers as `buf_len` resolves to `size_t`:

powerpc, mipsel, armhf and armel archs have the same issues.

anatoli26 commented:
And just found in `sieve/sieved.c`:

`static void print_time(uint64_t t)` - looks like it's unused, but if not, is it correct to use a hardcoded 64bit type for time variables?

anatoli26 commented:
@elliefm, I've just fpushed the fixes for `PRIu64` and hardcoded format specifier for `buf_len`, and added a new commit defining `RLIM_T_FMT` and `strtorlim` the same way you've done it for other types. Builds OK on OpenBSD.

@guimard, I guess Ellie is out-of-office now, could you please meanwhile test the new code of this PR just to be sure is has no known issues? A consolidated patch is available here: https://github.com/cyrusimap/cyrus-imapd/pull/3262.patch

guimard commented:
@anatoli26: I've this error:
```
configure: error: Unable to determine underlying integer type of `rlim_t'
```
Is there a missing patch ?

anatoli26 commented:
> Is there a missing patch ?

Nop. I've just tried it on Linux and got the same error, works well on OpenBSD though.

It looks like there's some include missing on Linux for `rlim_t`, probably `resources.h` that should be explicitly included in the type tests, but I'll leave it for @elliefm.

anatoli26 commented:
@guimard, I've just fpushed a fix. Now it compiles correctly on both OpenBSD and Linux (the correct include was `sys/resource.h`).

Please try to run it. A consolidated patch is available here: https://github.com/cyrusimap/cyrus-imapd/pull/3262.patch (which is the URL of this PR + `.patch` at the end)

guimard commented:
@anatoli26: build succeeded locally, I'm going to push to debian machines, result [here](https://buildd.debian.org/status/logs.php?pkg=cyrus-imapd&ver=3.2.4-6) in a few minutes (may be 1 hour)

anatoli26 commented:
@guimard, thanks!

@elliefm, now there are no warnings related to this PR on any arch of the build farm, though I would suggest checking the other warnings. Most of them are nothing special like the ones we've seen in the other issue and things like (non-)reproducible builds, but some of them, like pointer casts to int of different size, signedness, etc. + 64bit types and data structures on 32bit archs IMO should be checked with more attention.

Also my latest fix for `buf_len()` implies that the max size it can handle is less than 4Gb on 32bit platforms (`buf_len()` returns `size_t` which is 32bit). Is that an expected behavior?

guimard commented:
@elliefm: the last build embeds:
 * Debian patches _(like all my uploads, no changes)_:
   * [Disable runpath checking](https://salsa.debian.org/debian/cyrus-imapd/-/blob/master/debian/patches/0001-Disable-runpath-checking.patch)
   * [Shutdown and close sockets cleanly](https://salsa.debian.org/debian/cyrus-imapd/-/blob/master/debian/patches/0002-Shutdown-and-close-sockets-cleanly.patch)
   * [Fix syslog prefix _(missing "cyrus/")_](https://salsa.debian.org/debian/cyrus-imapd/-/blob/master/debian/patches/0003-Fix-syslog-prefix.patch)
   * [silence erroneous RLIMIT_NUMFDS-related log messages](https://salsa.debian.org/debian/cyrus-imapd/-/blob/master/debian/patches/0004-silence-erroneous-RLIMIT_NUMFDS-related-log-messages.patch)
   * Updates calling of the perl interpreter to what we expect in Debian
   * Fix paths for Debian in tools/rehash
   * [Adds some enhancements to sieveshell and fixes some paths](https://salsa.debian.org/debian/cyrus-imapd/-/blob/master/debian/patches/0007-Adds-some-enhancements-to-sieveshell-and-fixes-some-.patch)
   * [Accept invalid From: header in email](https://salsa.debian.org/debian/cyrus-imapd/-/blob/master/debian/patches/0008-Accept-invalid-From-header-in-email.patch)
   * [Normalize the authentication ID](https://salsa.debian.org/debian/cyrus-imapd/-/blob/master/debian/patches/0009-Normalize-the-authentication-ID.patch)
   * [Fix PATH_MAX on GNU/Hurd](https://salsa.debian.org/debian/cyrus-imapd/-/blob/master/debian/patches/0010-Fix-PATH_MAX-on-GNU-Hurd.patch)
   * [Fix extra libpci in SNMP_LIBS](https://salsa.debian.org/debian/cyrus-imapd/-/blob/master/debian/patches/0011-Fix-extra-libpci-in-SNMP_LIBS.patch)
   * Use UnicodeData.txt from system
   * [conversations: update the G keys even if the message has no threadid](https://salsa.debian.org/debian/cyrus-imapd/-/blob/master/debian/patches/0016-conversations-actually-need-to-update-GUID-counts-after-calculating-CID.patch)
   * Increase test timeout
 * HTTP/2 fix (2696ed7)
 * #3262 last patch

I'm Cyrus-Imapd maintainer in Debian Project for 1 year, I don't know if the Debian patches have been reported (probably not). I added links to patches that may interest you. I can build some MR with some of them if you want.

Cheers,
Xavier

anatoli26 commented:
@elliefm, there's small issue with the `rlim_t` size check. On all archs under Linux `./configure` shows: `checking size of rlim_t... 0` (and on OpenBSD it shows the correct size).

`config.log` shows:
```
677:configure:17531: checking size of rlim_t
680:conftest.c:69:57: error: 'rlim_t' undeclared (first use in this function)
```

The `AC_CHECK_SIZEOF(rlim_t)` should include `sys/resource.h` either explicitly in the 3rd param or implicitly via `AC_INCLUDES_DEFAULT`.

Anyway, it's a cosmetic issue at this stage.

elliefm commented:
> 64bit types and data structures on 32bit archs IMO should be checked with more attention.

These multi-platform tests are very useful for finding bad assumptions and stress-testing the configure script, but they also produce a LOT of noise.

For example, Cyrus is an email and calendar server, both of which features need precise timestamps into the future and past, and to store potentially a lot of data.  There is absolutely no reason for anyone to be running Cyrus in 2020 on a platform whose `time_t` or `size_t` is only 32 bits!  (We even warn about it from configure.)  So, it's useful to see what i386 et al complain about, but we don't really need to expend any effort making it portable to them if the only problem is portability.

Using the c99 fixed-size types (like uint64_t etc) is fine too -- it's completely reasonable to say that a particular variable needs to be a particular size and signedness (we often need this specificity when the data being represented is from one of our binary cyrusdb files).  If the platform doesn't support that size or signedness, we simply don't support the platform.

The cunit tests are all passing, so nothing looks too out of place in a way that matters (or at least, in way that matters and is tested for...)

> in the i386 build there are more than 300 warnings

How did you get this number?  I suspect it's polluted with the irrelevant library path warnings from libtool...

I checked the log by eye, and it looks like the interesting ones (`rlim_t`, `buf_len()`) you've already found and taken care of.  Great :)

elliefm commented:
> And just found in `sieve/sieved.c`:
> 
> `static void print_time(uint64_t t)` - looks like it's unused, 

It's a callback function for `print_vallist`.  It's used at line 828.

> but if not, is it correct to use a hardcoded 64bit type for time variables?

Yeah it's reasonable.  Any platform worth actually using has a 64-bit time_t these days (32 bit time_t overflows on dates beyond 2038...) .  

The fact that it's unsigned is strange (unsigned `time_t` can't represent dates before 1970), but I don't know this section of code, so I don't know if that's a real problem or an imaginary one in this context.  It's sieve though, which is mostly about email (not calendaring), and email didn't exist in 1970, so it probably doesn't matter...

elliefm commented:
@anatoli26 can you reorder your commits please?  At the moment the commit that uses RLIM_T_FMT occurs before the commit that detects and defines it, which will be a nuisance for anyone running a bisect past this code in the future.

@guimard That "fix path max on gnu/hurd" patch looks familiar... Ahh, f03b10ae1e.  We got a few patches from Debian around the same time, but I can't remember if they were from a GitHub issue/PR or from the mailing list, nor who submitted them.  Looks like the current version of the Debian patch is different from the one that's in our source, so I guess nobody removed the Debian patch after it was upstreamed, and it keeps getting local modifications...

"conversations: update the G keys..." looks like it was a Debian backport for 3.0.  It's 95d204ff12 and is already in 3.2, so you can just drop it.

"Normalize the authentication ID" is interesting.  It adds a new "normalizeuid" imapd.conf setting, and apparently it's been floating around a vendor-only patch since at least 2007, according to #964.  The apparent author, Jeroen, was a regular Cyrus contributor a few years ago, so it's strange that this is floating around as a vendor patch and not actually integrated??  Looks like there's a couple of other patches here from Jeroen that aren't upstream, very curious.

Some of the patches look like Debian-specific things, but some are probably useful more generally.  For the general stuff, I'm happy to shepherd through PRs to integrate them if you're happy to submit them. :)

elliefm commented:
> there's small issue with the `rlim_t` size check. On all archs under Linux `./configure` shows: `checking size of rlim_t... 0` (and on OpenBSD it shows the correct size).
> [...]
> Anyway, it's a cosmetic issue at this stage.

You can just remove that `AC_CHECK_SIZEOF(rlim_t)`: nothing uses the result, and it's not interesting enough to keep around for config.log

anatoli26 commented:
@elliefm,

> How did you get this number? I suspect it's polluted with the irrelevant library path warnings from libtool...

On the build results page [1] there are 308 warnings, most of them noise warnings. As I've already mentioned and as you replied, there are some 32bit vs 64bit and some signedness issues, but if the policy is not to support i386 and other 32bit archs, maybe it would be clearer and safer to explicitly say that in the docs and other places, so nobody even tries to build and run it on those archs and we save on maintenance?

BTW, on all archs there are these 2 warnings that I'm not sure about:
```
imap/global.c:355:24: warning: implicit conversion from ‘enum cyrus_opt’ to ‘enum imapopt’ [-Wenum-conversion]
  355 |       config_getswitch(CYRUSOPT_NORMALIZEUID));
      |                        ^~~~~~~~~~~~~~~~~~~~~
```
and
```
imap/message.c:314:39: warning: array subscript has type ‘char’ [-Wchar-subscripts]
  314 |   if (is_from > 0 && *p != from_header[is_from]) {
      |                                       ^
```

> @anatoli26 can you reorder your commits please? At the moment the commit that uses RLIM_T_FMT occurs before the commit that detects and defines it, which will be a nuisance for anyone running a bisect past this code in the future.

Done!

> You can just remove that `AC_CHECK_SIZEOF(rlim_t)`: nothing uses the result, and it's not interesting enough to keep around for config.log

Done!

Is there anything else pending for this PR?

[1] https://buildd.debian.org/status/fetch.php?pkg=cyrus-imapd&arch=i386&ver=3.2.4-6&stamp=1604675816&raw=0

guimard commented:
> "conversations: update the G keys..." looks like it was a Debian backport for 3.0. It's [95d204f](https://github.com/cyrusimap/cyrus-imapd/commit/95d204ff12daa16b1987e2ad1495eb5e73556c49) and is already in 3.2, so you can just drop it.

@elliefm: removed, thanks!

> "Normalize the authentication ID" is interesting. It adds a new "normalizeuid" imapd.conf setting, and apparently it's been floating around a vendor-only patch since at least 2007, according to #964. The apparent author, Jeroen, was a regular Cyrus contributor a few years ago, so it's strange that this is floating around as a vendor patch and not actually integrated?? Looks like there's a couple of other patches here from Jeroen that aren't upstream, very curious.
> 
> Some of the patches look like Debian-specific things, but some are probably useful more generally. For the general stuff, I'm happy to shepherd through PRs to integrate them if you're happy to submit them. :)

Done also :wink:

elliefm commented:
> BTW, on all archs there are these 2 warnings that I'm not sure about:
> 
> ```
> imap/global.c:355:24: warning: implicit conversion from ‘enum cyrus_opt’ to ‘enum imapopt’ [-Wenum-conversion]
>   355 |       config_getswitch(CYRUSOPT_NORMALIZEUID));
>       |                        ^~~~~~~~~~~~~~~~~~~~~
> ```

This is from the "Normalize the authentication ID" Debian patch, which @guimard has now submitted as #3283 (but I have not yet looked at further).  We can ignore it here, and deal with it there instead; it's not related to this PR and is not currently upstream either.

> 
> and
> 
> ```
> imap/message.c:314:39: warning: array subscript has type ‘char’ [-Wchar-subscripts]
>   314 |   if (is_from > 0 && *p != from_header[is_from]) {
>       |                                       ^
> ```

I haven't seen this warning before, but looks like it's pointing out that `char`'s signedness is platform dependent and could potentially be signed and negative...

But now I've searched master and 3.2 for `from_header` and it's not in upstream code, so I guess this is from another one of the Debian patches -- which would explain why I don't see the warning on my own builds!  In either case, it's not related to this PR.

> Is there anything else pending for this PR?

I don't think so, but I want to check it over again just in case.  I'll do that today :)

elliefm commented:
I've also merged this onto the 3.2 branch

anatoli26 commented:
Great! Thanks!

ksmurchison commented on cmulocal/cyr_inttype.m4:
This method of forcing a warning fails on my dev box using:

gcc (GCC) 10.2.1 20201016 (Red Hat 10.2.1-6)
CPPFLAGS="-W -Wall -Werror -Wno-cpp"

I have to remove -Werror in order for configure to complete successfully.  Since I always use -Werror to check for questionable code in Cyrus, this change is a non-starter for me.  Not to mention that it seems somewhat fragile.

ksmurchison replied:
I just committed https://github.com/cyrusimap/cyrus-imapd/commit/b3f30361953a915e40dab27689ff4daf6f35ae0b

Which works arounf the problem for me, pending a possible better fix

obache replied:
Such flags should not be used for configure, or some kind of feature detections may fail silently and will be disabled,
OK to use for building, though, just to check codes.

elliefm replied:
@ksmurchison Oh right, thanks. I already did this for CFLAGS (because I always have -Werror in CFLAGS), but forgot about CPPFLAGS.  I'll tidy this up to put the CPPFLAGS handling in the same place as the CFLAGS handling for consistency, but I'll also check docs and see if there's anything else that needs the same treatment.
